File: solutions/_473_matchsticks_to_square/solution.js
/**
 * @param {number[]} matchsticks
 * @return {boolean}
 */
var makesquare = function (matchsticks) {
  if (matchsticks.length < 4) {
    return false;
  }
  const perimeter = matchsticks.reduce((acc, curr) => acc + curr, 0);
  if (perimeter % 4 !== 0) {
    return false;
  }
  const side = perimeter / 4;
  matchsticks.sort((a, b) => b - a);
  const sides = [0, 0, 0, 0];
  return dfs(0);

  function dfs(i) {
    if (i === matchsticks.length) {
      return (
        sides[0] === side &&
        sides[1] === side &&
        sides[2] === side &&
        sides[3] === side
      );
    }

    for (let j = 0; j < 4; j++) {
      if (sides[j] + matchsticks[i] <= side) {
        sides[j] += matchsticks[i];
        if (dfs(i + 1)) {
          return true;
        }
        sides[j] -= matchsticks[i];
      }
    }
    return false;
  }
};

module.exports = { makesquare };

/*
https://leetcode.com/problems/matchsticks-to-square/
*/

File: solutions/_473_matchsticks_to_square/solution.py
from typing import List


class Solution:
    def makesquare(self, matchsticks: List[int]) -> bool:
        n = len(matchsticks)
        if n < 4:
            return False
        perimeter = sum(matchsticks)
        if perimeter % 4 != 0:
            return False
        side = perimeter // 4
        matchsticks.sort(reverse=True)
        sides = [0] * 4

        def dfs(i):
            if i == n:
                return sides[0] == sides[1] == sides[2] == side
            for j in range(4):
                if sides[j] + matchsticks[i] <= side:
                    sides[j] += matchsticks[i]
                    if dfs(i + 1):
                        return True
                    sides[j] -= matchsticks[i]
            return False

        return dfs(0)


# https://leetcode.com/problems/matchsticks-to-square/
